README.md
# Checkout POS for WooCommerce

**Contributors:** [Miika Arponen](https://github.com/nomafin), [Ville Siltala](https://github.com/villesiltala), [Tomi Henttinen](https://github.com/tomihenttinen), [Indre Solodov](https://github.com/Indre87) & [Joonas Loueranta](https://github.com/loueranta)

**Requires**
- [WordPress](https://wordpress.org/download/): 4.9.0 and up
- [WooCommerce](https://wordpress.org/plugins/woocommerce/): 3.0.0 and up
- PHP: 7.1 and up

## Description

Connect your [Checkout POS](https://www.checkout.fi/pos) (formerly OP Kassa) and WooCommerce to synchronize products, orders and stock levels between the systems.

## Installation

The plugin is published at the WordPress.org plugin directory to make installation and updates easier.

1. Log on to WordPress admin area and navigate to **Plugins** -> **Add New**.
2. Type "[Checkout POS for WooCommerce](https://wordpress.org/plugins/op-kassa-for-woocommerce/)" to search field.
3. Install and activate the plugin from search results.
4. Head over to **WooCommerce** -> **Settings** and select the "OP Kassa" tab to configure the plugin.

### System Audit

The Plugin has a System Audit feature which is ran on plugin activation and may be also ran manually from plugin settings page.

The System Audit checks for the following:

1. The system settings requirements are met ('limit'-value needs to be met or exceeded):
    1. `memory_limit`
    2. `max_execution_time`
2. WordPress-options are configured properly ('value'-value needs to match the Wordpress configuration):
    1. `permalink_structure`
    2. `woocommerce_calc_taxes` (warn only)
3. Mandatory plugins are installed/activated
4. Incompatible plugins are not installed (may issue an warning or error)
5. System has connection to target systems

If the System Audit fails or shows warnings, please contact Checkout POS support. And attach screenshot of the result with your message.

## Configuration

The plugin adds Checkout POS configuration tab to WooCommerce settings. 

The standard URL for this configuration tab is:
```
/wp-admin/admin.php?page=wc-settings&tab=kis
```

### Connecting to Checkout POS

On the Checkout POS settings tab, the user can activate connections required by Checkout POS.

Merchant details are found on the page after the connections are created successfully.

### Settings

There are currently two settings available for configuration on the Checkout POS tab in WooCommerce Settings: Product export direction and Order export direction.

For **Product export** you can choose to disable it (default setting) or choose which way you want the product data to be synchronized, from WooCommerce to Checkout POS or vice versa. 

For **Order export** you can choose to disable it (default setting) or choose to sychronize the order data both ways or just one way, from WooCommerce to Checkout POS or vice versa. 

Choosing the Stock export setting is currently disabled and is linked to Product export setting.

Please note that when you change the settings and hit save, it will take couple of minutes for the synchronization to start.

### QA/Test environment

If you are using OP Kassa QA environment, you need to select "Connect to Checkout POS Test Environment" on the settings tab. Please use this setting only if you know what you are doing as it will disable the connection to the production environment of Checkout POS.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.0.0] - 2021-06-08

### Changed
- New settings page
- Rebranding from "OP Kassa" to "Checkout POS"

## [2.0.1] - 2021-03-29

### Changed
- Added OP Kassa integration authentication related settings.

## [2.0.0] - 2021-03-23

### Changed
- Replaced OAuth based authentication with the WooCommerce Rest API authentication.

## [1.0.6] - 2021-02-05

### Changed
- It is now possible to choose whether Woo tax calculation is used on orders synced from OP Kassa or if OP Kassa tax calculation is used instead.

## [1.0.5] - 2020-11-23

### Changed
- OP Kassa is now disconnected gracefully if the Woo instance domain is changed.

## [1.0.4] - 2020-11-04

### Changed
- Fixed a bug relating to Kassa oauth callback url 

## [1.0.3] - 2020-10-27

### Changed
- Updated installation instructions

## [1.0.2] - 2020-10-27

### Added
- Released to Wordpress.org directory

## [1.0.1] - 2020-10-21

### Changed
- Changed all direct curl-usages to Wordpress HTTP API-calls.
- Added sanitation for all POST/GET/SERVER-inputs.

## [1.0.0] - 2020-10-05

### Changed
- Prepared the plugin for publishing.
- Added more descriptive error message for failed Kassa connection.

## [0.8.1] - 2020-09-16

### Changed
- Added a custom system check for Polylang-plugin setting.
- Changed the method to get the site domain name. Old way had issues with url mismatches with server-variable and the url defined on Wordpress settings.
- Call kassa-oauth-delete-function to remove the sync connection entry for the domain on Kassa disconnect.

## [0.8.0] - 2020-09-11

### Changed
- Added possibility to select the target Kassa environment (test/production).
- Kassa connection is now disconnected on plugin deactivation.
- Fixed calls to wp_timezone(), which threw an error on older (< WP 5.3) WordPress versions.

## [0.7.11] - 2020-07-02

### Changed
- Prevent putting orders to 'completed'-status directly from 'pending'-status if they are Kassa-orders created via API.

## [0.7.10] - 2020-06-17

### Changed
- Disabled Stock sync direction selection. Stock sync direction is set the same as for the Product sync.

## [0.7.9] - 2020-06-01

### Changed
- Disabled all order emails for orders which are created via Rest API.

## [0.7.8] - 2020-04-15

### Changed
- Added more information on mandatory or incompatible plugins to plugin's system audit result.

## [0.7.7] - 2020-04-09

### Changed
- Fixed an issue with custom date_query timestamps

## [0.7.6] - 2020-03-25

### Changed
- Fixed issue in memory_limit-check.

## [0.7.5] - 2020-03-20

### Changed
- Fixed typo from production default URLs.

## [0.7.4] - 2020-03-20

### Changed
- Updated production default URLs.

## [0.7.3] - 2020-03-19

### Changed
- Removed redundant duplicate system audit check for a mandatory plugin.

## [0.7.2] - 2020-03-19

### Changed
- Minor bug fix
- Renamed plugin basename

## [0.7.1] - 2020-03-18

### Changed
- Replaced the system audit configuration file url to point to production environment.

## [0.7] - 2020-03-13

### Changed
- Added system audit to perform basic enviroment checks

## [0.6.0] - 2019-10-29

### Changed
- Authentication occurs on a single request chain taking the user first to Kassa and then to WooCommerce authentication.
- Kassa and WooCommerce authentications are both canceled with a single action on the OP Kassa settings page.
- Webhook secret handling updated.

## [0.5.10] - 2019-15-10

### Removed
- Removed checks to ensure only the user that created OAuth link can remove it
- Removed filters to force EAN to be a number


## [0.5.9] - 2019-27-09

### Added
- Added production endpoint urls

### Fixed
- Fixed production endpoint API paths
- Fixed a typo in the plugin name

### Removed
- Removed "Initial stock sync" -setting from the settings page since it's not used


## [0.5.8] - 2019-23-09

### Added
- Added production endpoint urls

### Removed
- Removed custom REST processing for the Kassa EAN. Will use normal metadata instead


## [0.5.7] - 2019-18-09

### Fixed

- Fixed product EAN ajax processing in while editing product variants

## [0.5.6] - 2019-18-09

### Fixed

- Fixed handling for the `kis_modified_after` query parameter if the value is `"0"` _(meaning: starting from the beggining)_.

## [0.5.5] - 2019-18-09

### Fixed

- Fixed using `get_id()` only for objects from API requests

## [0.5.4] - 2019-18-09

### Fixed

- Fixed syntax error where $metaKeyName was not defined inside `update_callback` function in ProductEAN class
- Fixed classes using `WP_Post` ID-property directly instead of using `get_id()`

## [0.5.3] - 2019-18-09


src/Merchant.php
<?php
/**
 * This class is used to wrap the Checkout POS OAuth response data.
 */

namespace CheckoutFinland\WooCommerceKIS;

/**
 * Class Merchant
 *
 * This class is used to wrap the Checkout POS OAuth response data.
 *
 * @package CheckoutFinland\WooCommerceKIS
 */
class Merchant {

    /**
     * The merchant id.
     *
     * @var int
     */
    protected $merchant_id;

    /**
     * The merchant owner email.
     *
     * @var string
     */
    protected $owner_email;

    /**
     * The merchant mode.
     *
     * @var string
     */
    protected $merchant_mode;

    /**
     * The merchant owner phone number.
     *
     * @var string
     */
    protected $owner_phone_number;

    /**
     * * The merchant name.
     *
     * @var string
     */
    protected $merchant_name;

    /**
     * The merchant billing street name.
     *
     * @var string
     */
    protected $merchant_billing_street_name;

    /**
     * The merchant billing zip.
     *
     * @var string
     */
    protected $merchant_billing_zip;

    /**
     * The merchant billing city.
     *
     * @var string
     */
    protected $merchant_billing_city;

    /**
     * MerchantDetails constructor.
     *
     * @param object $data The decoded JSON data.
     */
    public function __construct( $data ) {
        if ( empty( $data ) ) {
            return;
        }

        foreach ( $data as $key => $value ) {
            if ( property_exists( $this, $key ) ) {
                $this->{$key} = $value;
            }
        }
    }

    /**
     * Get the merchant id.
     *
     * @return int
     */
    public function get_merchant_id() : ?int {

        return $this->merchant_id;
    }

    /**
     * Get the owner email.
     *
     * @return string
     */
    public function get_owner_email() : ?string {

        return $this->owner_email;
    }

    /**
     * Get the merchant mode.
     *
     * @return string
     */
    public function get_merchant_mode() : ?string {

        return $this->merchant_mode;
    }

    /**
     * Get the owner phone number.
     *
     * @return string
     */
    public function get_owner_phone_number() : ?string {

        return $this->owner_phone_number;
    }

    /**
     * Get the merchant_name.
     *
     * @return string
     */
    public function get_merchant_name() : ?string {

        return $this->merchant_name;
    }

    /**
     * Get the merchant billing street name.
     *
     * @return string
     */
    public function get_merchant_billing_street_name() : ?string {

        return $this->merchant_billing_street_name;
    }

    /**
     * Get the merchant billing zip.
     *
     * @return string
     */
    public function get_merchant_billing_zip() : ?string {

        return $this->merchant_billing_zip;
    }

    /**
     * Get the merchant billing city.
     *
     * @return string
     */
    public function get_merchant_billing_city() : ?string {

        return $this->merchant_billing_city;
    }

    /**
     * Format the merchant address into a printable format.
     *
     * @return string
     */
    public function format_address() : string {
        $address = [
            $this->merchant_billing_street_name,
            $this->merchant_billing_zip,
            $this->merchant_billing_city,
        ];

        $address = array_filter( $address );

        return implode( ', ', $address );
    }

    /**
     * Evaluate the object validity.
     *
     * @return bool
     */
    public function is_valid() {
        if ( empty( $this->owner_email ) ) {
            return false;
        }
        return true;
    }
}


plugin.php
<?php
/**
* Plugin Name: Checkout POS for WooCommerce
* Plugin URI: https://github.com/CheckoutFinland/checkout-pos-for-woocommerce 
* Description: Connect your Checkout POS (formerly OP Kassa) and WooCommerce to synchronize products, orders and stock levels between the systems.
* Version: 3.0.0
* Requires at least: 4.9
* Tested up to: 5.7
* Requires PHP: 7.1
* WC requires at least: 3.0
* WC tested up to: 5.3
* Author: Checkout Finland
* Author URI: https://www.checkout.fi/pos 
* Text Domain: op-kassa-for-woocommerce
* Domain Path: /languages
* License: MIT
* License URI: https://opensource.org/licenses/MIT
* Copyright: Checkout Finland
*/

namespace CheckoutFinland\WooCommerceKIS;

use CheckoutFinland\WooCommerceKIS\Activation;
use CheckoutFinland\WooCommerceKIS\Admin\Notice;
use CheckoutFinland\WooCommerceKIS\Admin\ProductEAN;
use CheckoutFinland\WooCommerceKIS\Admin\PackageSlip;
use CheckoutFinland\WooCommerceKIS\Admin\SettingsPage;
use CheckoutFinland\WooCommerceKIS\Admin\OrderInfoMetaBox;
use CheckoutFinland\WooCommerceKIS\Admin\DeletedProductTracker;

if ( ! defined( 'ABSPATH' ) ) {
    die;
}

if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
	require_once( __DIR__ . '/vendor/autoload.php' );
}

/**
 * Initializes plugin functionalities.
 *
 * @since      1.0.0
 * @package    CheckoutFinland\WooCommerceKIS
 */
final class Plugin {

    /**
     * Holds the plugin singleton.
     *
     * @since    0.0.0
     * @access   private
     * @var Plugin
     */
    private static $instance;

    /**
     * The notice instance handles displaying all admin notices.
     *
     * @since    0.0.0
     * @access   private
     * @var      Notice    $notice    Handles all admin notices.
     */
    protected $notice;

    /**
     * The WooCommerce REST API modifications class.
     *
     * @var Api
     */
    protected $api;

    /**
     * Order info box renderer for orders
     *
     * @since 0.4.0
     * @var OrderInfoMetaBox
     */
    protected $order_box;

    /**
     * Handler for product EAN related stuff
     *
     * @since 0.5.0
     * @var ProductEAN
     */
    protected $product_ean;

    /**
     * Package slip metabox render class
     *
     * @since 0.5.0
     * @var PackageSlip
     */
    protected $package_slip;

    /**
     * Tracker for deleted and trashed posts
     *
     * @since 0.4.0
     * @var DeletedProductTracker
     */
    protected $deleted_product_tracker;

    /**
     * Domain used to connect to OP Kassa
     *
     * @since 1.0.5
     * @var string
     */
    protected $kassa_connected_site_domain;

    /**
     * Define all WooCommerce post types here
     * since WC does not bother to define them.
     */
    const WC_POST_TYPES = [
        'product',
        'product_variation',
        'product_visibility',
        'shop_order',
        'shop_coupon',
        'shop_webhook',
    ];

    /**
     * Private constructor for the plugin singleton.
     */
    public function init() {
        $this->set_constants();

        $this->notice                  = new Notice();
        $this->api                     = new Api();
        $this->order_box               = new OrderInfoMetaBox();
        $this->deleted_product_tracker = new DeletedProductTracker();
        $this->product_ean             = new ProductEAN();
        $this->package_slip            = new PackageSlip();

        $this->set_locale();
        $this->define_admin_hooks();
        $this->define_global_hooks();
    }

    /**
     * Force disconnect from Kassa if site domain has changed
     * 
     * @since    1.0.5
     * @access   private
     */
    function check_siteurl_changes() {
        $site_url = parse_url( get_option( 'siteurl' ) );
        $kassa_connected_site_url = get_option( 'kassa_connected_site_url' );

        if ( !$kassa_connected_site_url ) {
            return;
        }

        $kassa_connected_site_url = parse_url( $kassa_connected_site_url );
        if( $kassa_connected_site_url['host'] && $kassa_connected_site_url['host'] !== $site_url['host'] ) {
            $this->kassa_connected_site_domain = $kassa_connected_site_url['host'];

            // Disconnect from Kassa
            $this->disconnect_kassa();
            delete_option( 'kassa_connected_site_url', null );

            // Add notification about the disconnect and about the possibility of double products/orders in Kassa after reconnect
            add_action( 'admin_notices', function () {
                $site_url = parse_url(get_option('siteurl'));
                ?>
                <div class="notice notice-warning is-dismissible">
                    <p>
                        <div><?php echo esc_html( 'The site domain has changed: OP Kassa has been disconnected.', 'woocommerce-kis' ); ?></div>
                        <div><?php echo esc_html( 'Old domain: ', 'woocommerce-kis' ) . $this->kassa_connected_site_domain; ?></div>
                        <div><?php echo esc_html( 'New domain: ', 'woocommerce-kis' ) . $site_url['host']; ?></div>
                        <div><?php echo esc_html( 'Please make sure the new domain is correct before re-connecting to OP Kassa.', 'woocommerce-kis' ); ?></div>
                        <div><?php echo esc_html( 'NOTE! If You connect to OP Kassa with same credentials than earlier, the products may be duplicated during the product synchronization. You may want to delete the products from OP Kassa prior new connection.', 'woocommerce-kis' ); ?></div>
                    </p>
                </div>
                <?php
            });
        }
    }

    /**
     * Define plugin constants.
     */
    private function set_constants() {
        $plugin_headers = [
            'Version' => 'Version',
        ];

        if ( function_exists( 'get_file_data' ) ) {
            $plugin_data = \get_file_data( __FILE__, $plugin_headers, 'plugin' );

            $kis_version = $plugin_data['Version'];
        }
        else {
            $kis_version = 'unknown';
        }

        define( 'WOOCOMMERCE_KIS_VERSION', $kis_version );

        if ( ! defined('KIS_HAS_CUSTOM_ENVIRONMENT')) {
            define( 'KIS_HAS_CUSTOM_ENVIRONMENT', $this->has_custom_environment_urls());
        }

        $kis_test_environment_enabled = get_option('kis_test_environment_enabled', null);
        // If kis_test_environment_enabled-option has not been set yet, use 'no' as default value
        if (is_null($kis_test_environment_enabled)) {
            $kis_test_environment_enabled = 'no';
        }

        if ( ! defined( 'KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED' ) ) {
            define( 'KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED', $kis_test_environment_enabled );
        }

        if ( ! defined( 'KIS_WOOCOMMERCE_OAUTH_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ? 
                'https://woocommerce.prod.op-kassa.fi/prod/woo-auth-initiate' : 
                'https://woocommerce.qa.op-kassa.fi/qa/woo-auth-initiate';
            define( 'KIS_WOOCOMMERCE_OAUTH_URL', $target_url );
        }

        if ( ! defined( 'KIS_KASSA_OAUTH_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ?
                'https://woocommerce.prod.op-kassa.fi/prod/kassa-oauth-initiate' :
                'https://woocommerce.qa.op-kassa.fi/qa/kassa-oauth-initiate';
            define( 'KIS_KASSA_OAUTH_URL', $target_url );
        }

        if ( ! defined( 'KIS_KASSA_DELETE_OAUTH_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ?
                'https://woocommerce.prod.op-kassa.fi/prod/kassa-oauth-delete' :
                'https://woocommerce.qa.op-kassa.fi/qa/kassa-oauth-delete';
            define( 'KIS_KASSA_DELETE_OAUTH_URL', $target_url );
        }

        if ( ! defined( 'KIS_WOOCOMMERCE_OAUTH_CALLBACK_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ?
                'https://woocommerce.prod.op-kassa.fi/prod/woo-auth-callback' :
                'https://woocommerce.qa.op-kassa.fi/qa/woo-auth-callback';
            define( 'KIS_WOOCOMMERCE_OAUTH_CALLBACK_URL', $target_url );
        }

        if ( ! defined( 'KIS_WOOCOMMERCE_WEBHOOK_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ?
                'https://woocommerce.prod.op-kassa.fi/prod/woo-webhook' :
                'https://woocommerce.qa.op-kassa.fi/qa/woo-webhook';
            define( 'KIS_WOOCOMMERCE_WEBHOOK_URL', $target_url );
        }

        if ( ! defined( 'KIS_WOOCOMMERCE_SYSTEM_AUDIT_CONFIG_URL' ) ) {
            $target_url = KIS_WOOCOMMERCE_TEST_ENVIRONMENT_ENABLED === 'no' ?
                'https://woocommerce.prod.op-kassa.fi/prod/wooClientSystemAuditConfig.json' :
                'https://woocommerce.qa.op-kassa.fi/qa/wooClientSystemAuditConfig.json';
            define( 'KIS_WOOCOMMERCE_SYSTEM_AUDIT_CONFIG_URL', $target_url );
        }
    }

    /**
     * Returns true if some of the environmental urls are set
     * 
     * @return bool
     * @since    0.8.0
     * @access   private
     */
    private function has_custom_environment_urls() : bool {
        return (defined( 'KIS_WOOCOMMERCE_OAUTH_URL' ) ||
            defined( 'KIS_KASSA_OAUTH_URL' ) ||
            defined( 'KIS_WOOCOMMERCE_OAUTH_CALLBACK_URL' ) || 
            defined( 'KIS_WOOCOMMERCE_WEBHOOK_URL' ) ) ;
    }

    /**
     * Disconnects the plugin from Kassa: 
     * - if plugin is connected to Kassa but the target environment is not known
     * 
     * @since    0.8.0
     * @access   private
     */
    private function disconnect_from_kassa_if_no_environment() {
        $kis_test_environment_enabled = get_option('kis_test_environment_enabled', null);

        if (is_null($kis_test_environment_enabled)) {
            $this->disconnect_kassa();
            // Add the kis_test_environment_enabled-option with a default value of 'no'
            add_option('kis_test_environment_enabled', 'no');
        }
    }

    /**
     * If given new option value differs from the previous option value:
     * - Disconnect from Kassa
     * - Reload plugin to redefine new environment related constants
     * 
     * @since    0.8.0
     * @access   public
     */
    public function reload_plugin( $new_option_value, $old_option_value ) {
        if ( $new_option_value !== $old_option_value && ! empty( $new_option_value ) ) {
            $this->disconnect_kassa();

            // Reload page
            header("Refresh:0");
        }
    }

    /**
     * Force disconnect from Kassa if connection is active
     * 
     * @return bool disconnect was made
     * 
     * @since    0.8.0
     * @access   private
     */
    private function disconnect_kassa() : bool {
        $oauth = new OAuth( $this->notice );

        if ( $oauth->is_oauth_active() ) {
            $this->oauth_delete();
            $oauth->handle_oauth_delete(true);

            return true;
        }

        return false;
    }

    /**
     * Send 'Disconnect from Kassa'-request to the integration
     * 
     * @return bool disconnect was made
     * 
     * @since    0.8.1
     * @access   private
     */
    private function oauth_delete() : bool {
        $url = Utility::add_query_parameter( KIS_KASSA_DELETE_OAUTH_URL, 'domain', Utility::get_server_name() );

        $http_code = wp_remote_retrieve_response_code( wp_remote_get( $url ) );

        if ( $http_code >= 400) {
            error_log('Kassa Disconnect failed! Check OP Kassa integration logs for further information.');
        }

        return $http_code < 400;
    }

    /**
     * Define the locale for this plugin for internationalization.
     *
     * Uses the Plugin_Name_i18n class in order to set the domain and to register the hook
     * with WordPress.
     *
     * @since    0.0.0
     * @access   private
     */
    private function set_locale() {
        $plugin_i18n = new I18n();

        add_action( 'plugins_loaded', [ $plugin_i18n, 'load_plugin_textdomain' ] );
    }

    /**
     * Register all of the hooks related to the admin area functionality
     * of the plugin.
     *
     * @since    0.0.0
     * @access   private
     */
    private function define_admin_hooks() {

        // Hooks to handle plugin environment change
        add_filter( 'update_option_kis_test_environment_enabled', [$this, 'reload_plugin'], 10, 2 );

        // Hooks to handle plugin auth method change
        add_filter( 'update_option_kis_woo_auth_params_enabled', [$this, 'reload_plugin'], 10, 2 );


        // Register the settings page for WooCommerce.
        add_filter(
            'woocommerce_get_settings_pages', [ SettingsPage::class, 'include_settings_page' ], 1, 1
        );

        // Hook for displaying OAUTH errors.
        add_filter(
            'admin_notices', [ $this->notice, 'oauth_error' ]
        );

        // Create a new metabox for orders
        add_action('add_meta_boxes', [$this->order_box, 'orderbox_metabox']);

        // Hook to add order metadata to order
        add_action('post_updated', [$this->order_box, 'save_metabox_data'], 10, 3);

        // Create a new metabox for package slip
        add_action('add_meta_boxes', [$this->package_slip, 'render_metabox']);

        // To track if posts are deleted, trashed or untrashed
        add_action('delete_post', [$this->deleted_product_tracker, 'post_deleted_or_trashed'], 10, 1);
        add_action('wp_trash_post', [$this->deleted_product_tracker, 'post_deleted_or_trashed'], 10, 1);
        add_action('untrash_post', [$this->deleted_product_tracker, 'post_untrashed'], 10, 1);

        // WooCommerce refund hook to update the order
        add_action('woocommerce_order_refunded', [$this, 'refresh_order_on_refund'], 10, 2 );

        // WooCommerce products EAN hooks
        add_action('woocommerce_product_options_sku', [$this->product_ean, 'add_ean_input_to_product']);
        add_action('woocommerce_process_product_meta', [$this->product_ean, 'save_product_ean']);

        // WooCommerce product variation EAN hooks
        add_action('woocommerce_variation_options', [$this->product_ean, 'add_ean_input_to_variation'], 10, 3);
        add_action('woocommerce_save_product_variation', [$this->product_ean, 'save_variation_ean'], 10, 2);
    }

    /**
     * Register all hooks run globally.
     *
     * @since    0.0.0
     * @access   private
     */
    private function define_global_hooks() {
        $notice = $this->notice;

        // Hook to the OAuth server plugin's token data hook.
        $oauth = new OAuth( $notice );
        add_filter(
            'json_oauth1_access_token_data', [ $oauth, 'handle_json_access_token_data' ], 1, 1
        );

        add_action( 'init', [ $this, 'check_siteurl_changes' ] );
        add_action( 'admin_init', [ $oauth, 'handle_kassa_oauth_response' ] );

        add_action( 'wp_loaded', function() {
            $this->disconnect_from_kassa_if_no_environment();
        });

        // Filter WooCommerce REST API query for all WC post types.
        $post_types = static::WC_POST_TYPES;
        array_walk(
            $post_types, function( $post_type ) {
                add_filter(
                    "woocommerce_rest_{$post_type}_object_query",
                    [ $this->api, 'filter_wc_rest_query' ],
                    PHP_INT_MAX,
                    2
                );
            }
        );

        // Add a custom route for the deleted products
        add_action(
            'rest_api_init',
            function () {
                register_rest_route(
                    'wc/v3',
                    'deleted_products',
                    [
                        'methods'  => 'GET',
                        'callback' => [$this->deleted_product_tracker, 'get_deleted_products_callback'],
                        'permission_callback' => function() {
                            return \current_user_can('manage_woocommerce');
                        }
                    ]
                );
                register_rest_route(
                    'wc/v3',
                    'delete_auth_token',
                    [
                        'methods'  => 'GET',
                        'callback' => [$this->api, 'delete_auth_token'],
                        'permission_callback' => function() {
                            return \current_user_can('manage_woocommerce');
                        }
                    ]
                );
            }
        );

        // Prevent putting orders to 'completed'-status directly from 'pending'-status if they are Kassa-orders created via API.
        // Auto completing of orders happens when no products on the order can be found from Woo.
        add_action( 'woocommerce_order_status_changed', function ( $order_id, $old_status, $new_status, $instance ) {
            $order = wc_get_order($order_id);
            if ( $order && $order->get_created_via() === 'rest-api' && $old_status === 'pending' && $new_status === 'completed' ) {
                // Can we trust this meta attribute is always on the order?
                $is_kassa_purchase = get_post_meta( $order_id, '_kassa_purchase', true );
                if ( $is_kassa_purchase ) {
                    $order->update_status( 'processing' );
                }
            }
        }, 10, 4 );

        // Prevent sending order emails for orders which are created via API
        add_action('woocommerce_new_order', function ($order_id) {         
             $order = wc_get_order($order_id);
         
             if ($order && $order->get_created_via() === 'rest-api') {
               add_filter( 'woocommerce_email_enabled_new_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_cancelled_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_failed_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_on_hold_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_processing_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_completed_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_refunded_order', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_invoice', function($yesno, $object){
                 return false;
               }, 10, 2);
               add_filter( 'woocommerce_email_enabled_customer_note', function($yesno, $object){
                 return false;
               }, 10, 2);
             }
         }, 10, 1);

         /**
         * Register the plugin deactivation hook.
         */
        register_deactivation_hook( __FILE__, array ( $this, 'plugin_deactivate' ));
    }

    /**
     * Check if the 'WP REST API - OAuth 1.0a Server' plugin is enabled.
     *
     * @see https://wordpress.org/plugins/rest-api-oauth1/
     *
     * @return bool
     */
    private function oauth_enabled() {
        return class_exists( 'WP_REST_OAuth1' );
    }

    /**
     * Get the plugin assets url.
     *
     * @return string
     */
    public function get_assets_url() : string {
        return plugin_dir_url( __FILE__ ) . 'assets/';
    }

    /**
     * Get the plugin assets path.
     *
     * @return string
     */
    public function get_assets_path() : string {
        return plugin_dir_path( __FILE__ ) . 'assets';
    }

    /**
     * Initializes the plugin once and returns the instance.
     */
    public static function instance() {
        if ( empty( static::$instance ) ) {
            static::$instance = new Plugin();

            return static::$instance;
        }

        return static::$instance;
    }

    /**
     * Updates order post modified timestamp when order is refunded
     *
     * @param int $order_id
     * @param int $refund_id
     * @return void
     */
    public function refresh_order_on_refund($order_id, $refund_id) {
        $order = get_post($order_id);
        $err = null;
        wp_update_post($order, $err);
    }

    /**
     * Run when the plugin is deactivated.
     * 
     * @since    0.8.0
     * @access   public
     */
    public function plugin_deactivate() {
        // Disconnect from Kassa on plugin deactivation.
        $this->disconnect_kassa();
    }
}

/**
 * A global method for getting the plugin singleton.
 *
 * @package CheckoutFinland\WooCommerceKIS
 * @since   0.0.0
 * @return  Plugin
 */
function plugin() {
    return Plugin::instance();
}

// Begin plugin excecution by creating the singleton.
plugin()->init();

/**
 * Register an Activation hook for handling the plugin system audit
 */
include_once dirname( __FILE__ ) . '/src/Admin/SystemAudit.php';
register_activation_hook( __FILE__, array( 'CheckoutFinland\WooCommerceKIS\Admin\SystemAudit', 'perform_system_audit' ) );
add_action( 'admin_notices', array( 'CheckoutFinland\WooCommerceKIS\Admin\SystemAudit', 'display_system_audit_admin_notice' ) );
